src/algorithms/async-model-average.md

```diff
@@ -0,0 +1,57 @@
+# Asynchronous Model Average
+
+In synchronous communication algorithms such as [Gradient AllReduce](./gradient-allreduce.md),
+every worker needs to be in the same iteration in a lock-step style. When there is no straggler
+in the system, such synchronous algorithms are fast, and often gives deterministic training results
+that are easier to reason about. However, when there are stragglers in the system, synchronous algorithms
+require faster workers to wait for the slowest one in each iteration, which can dramatically reduce
+the overall performance of the whole system. To solve this problem, we can use an asynchronous communication 
+algorithm that does not require every worker to be in the same iteration.
+
+The *Asynchronous Model Average* algorithm provided by Bagua is one of such algorithms.
+
+## Algorithm
+
+The *Asynchronous Model Average* algorithm can be described as follows: 
+
+Every worker maintains a local model ${\bf x}$. The $i$-th worker maintains $\mathbf{x}_i$.
+Every worker repeats the following three steps, for the $i$-th worker:
+
+1. **Compute gradients**: Calculate a local gradient $\nabla F({\bf x}^{(i)})$, where $\hat {\bf x}^{(i)}$ is a clone of $\mathbf{x}_i$.
+2. **Gradient update**: Update the model with ${\bf x}^{(i)} = {\bf x}^{(i)} - \gamma \nabla F(\hat {\bf x}^{(i)}) $, note that $\hat {\bf x}^{(i)}$ may not be the
+same with ${\bf x}^{(i)}$ as it could be modified in the **averaging** step.
+3. **Average**: Average local model ${\bf x}^{(i)}$ with all other workers' models. More specifically,
+ $ {\bf x}^{(i)} = \frac{1}{n} \sum_{i'=1}^{n} {\bf x}^{(i')} $.
+
+All workers run the procedure above simultaneously.
+
+**Gradient update** and **averaging** can run in parallel and the **averaging** step needs to be atomic.
```

Comment:
This is so hard to understand. Give more detail and structure, for example describe it as two separate threads, a computation thread that continuously computes gradients, and a communication thread that continuously does averaging and weights update. See Appendix A in https://arxiv.org/pdf/1710.06952.pdf for an example.